File: 2025/ProjectManagement/backend/src/main/java/com/projectmanagement/service/ProjectService.java
package com.projectmanagement.service;

import com.projectmanagement.model.Project;
import com.projectmanagement.repository.ProjectRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ProjectService {

    @Autowired
    private ProjectRepository projectRepository;

    public List<Project> getAllProjects() {
        return projectRepository.findAll();
    }

    public Project getProjectById(Long id) {
        Optional<Project> project = projectRepository.findById(id);
        return project.orElse(null);
    }

    public List<Project> searchProjects(String query) {
        return projectRepository.searchProjects(query);
    }

    public List<Project> getProjectsByStatus(Project.Status status) {
        return projectRepository.findByStatus(status);
    }

    public List<Project> getProjectsByPriority(Project.Priority priority) {
        return projectRepository.findByPriority(priority);
    }

    public List<Project> getProjectsByCategory(Project.Category category) {
        return projectRepository.findByCategory(category);
    }

    public Project createProject(Project project) {
        if (project.getProgress() == null) {
            project.setProgress(0);
        }
        return projectRepository.save(project);
    }

    public Project updateProject(Long id, Project projectDetails) {
        Optional<Project> optionalProject = projectRepository.findById(id);
        if (optionalProject.isPresent()) {
            Project project = optionalProject.get();
            project.setName(projectDetails.getName());
            project.setDescription(projectDetails.getDescription());
            project.setStatus(projectDetails.getStatus());
            project.setPriority(projectDetails.getPriority());
            project.setCategory(projectDetails.getCategory());
            project.setStartDate(projectDetails.getStartDate());
            project.setEndDate(projectDetails.getEndDate());
            project.setProgress(projectDetails.getProgress() != null ? projectDetails.getProgress() : 0);
            project.setBudget(projectDetails.getBudget());
            project.setTeamMembers(projectDetails.getTeamMembers());
            project.setProjectManager(projectDetails.getProjectManager());
            project.setNotes(projectDetails.getNotes());
            return projectRepository.save(project);
        }
        return null;
    }

    public boolean deleteProject(Long id) {
        if (projectRepository.existsById(id)) {
            projectRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: 2025/ProjectManagement/frontend/src/app/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Project } from '../models/project';

@Injectable({
  providedIn: 'root'
})
export class ProjectService {
  private apiUrl = 'http://localhost:8100/api/projects';

  constructor(private http: HttpClient) {}

  getAllProjects(): Observable<Project[]> {
    return this.http.get<Project[]>(this.apiUrl);
  }

  getProjectById(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.apiUrl}/${id}`);
  }

  searchProjects(query: string): Observable<Project[]> {
    const params = new HttpParams().set('query', query);
    return this.http.get<Project[]>(`${this.apiUrl}/search`, { params });
  }

  getProjectsByStatus(status: string): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/status/${status}`);
  }

  getProjectsByPriority(priority: string): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/priority/${priority}`);
  }

  getProjectsByCategory(category: string): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/category/${category}`);
  }

  createProject(project: Project): Observable<Project> {
    return this.http.post<Project>(this.apiUrl, project);
  }

  updateProject(id: number, project: Project): Observable<Project> {
    return this.http.put<Project>(`${this.apiUrl}/${id}`, project);
  }

  deleteProject(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

